singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,377 @@
+    printf("testing BP0\n");
+    EOS_ERROR("Required Vinet model parameter BP0 < 1");
+    return;
+  }
+  if (_A0 < 0.0) {
+    printf("testing A0\n");
+    EOS_ERROR("Required Vinet model parameter A0 < 0");
+    return;
+  }
+  if (_Cv0 < 0.0) {
+    printf("testing Cv0\n");
+    EOS_ERROR("Required Vinet model parameter Cv0 < 0");
+    return;
+  }
+}
+
+PORTABLE_INLINE_FUNCTION void Vinet::InitializeVinet(const Real *d2tod40input) {
+
+  // The PressureCoeffsd2tod40Size (=39) allowed d2 to d40 coefficients
+  // for the pressure reference curve vs rho
+  // are seldome all used so did not want to crowd the argument list with them.
+  // Instead I ask the host code to send me a pointer to this array so that I can
+  // copy it here. Not used coeffs should be set to 0.0 (of course).
+  for (int ind = 0; ind < PressureCoeffsd2tod40Size; ind++) {
+    _d2tod40[ind] = d2tod40input[ind];
+  }
+  // Put a couple of  much used model parameter combinations and
+  // the energy coefficients f0 to f40 in an internal parameters array
+  // of size 2+41=VinetInternalParametersSize.
+  _VIP[0] = _B0 / _rho0 + pow(_A0 * _B0 / _rho0, 2.0) * _T0 / _Cv0; // sound speed squared
+  _VIP[1] = 3.0 / 2.0 * (_BP0 - 1.0);                               // exponent eta0
+  // initializing with pressure coeffs to get energy coeffs into VIP
+  _VIP[2] = 1.0;                                                // prefactor d0
+  _VIP[3] = 0.0;                                                // prefactor d1
+  for (int ind = 4; ind < VinetInternalParametersSize; ind++) { //_d2tod40[0]=d2
+    _VIP[ind] = _d2tod40[ind - 4];                              // dn is in VIP[n+2]
+  }
+  for (int ind = VinetInternalParametersSize - 2; ind >= 2;
+       ind--) { // _VIP[42]=d40=f40 given,first calculated is _VIP[41]=f39
+    _VIP[ind] = _VIP[ind] - (ind) / _VIP[1] * _VIP[ind + 1]; // prefactors f40 to f0
+  }                                                          // _VIP[n+2]=fn, ind=n+2
+}
+
+PORTABLE_INLINE_FUNCTION void Vinet::Vinet_F_DT_func(const Real rho, const Real T,
+                                                     Real *output) const {
+  int pref0vp = -2, pref0vip = 2, maxind = VinetInternalParametersSize - 3;
+  Real sumP = 0.0, sumB = 0.0, sumE = 0.0;
+
+  Real x = pow(_rho0 / rho, 1.0 / 3.0); /*rho-dependent*/
+  Real x2inv = 1.0 / pow(x, 2.0);
+  Real onemx = 1.0 - x;
+  Real etatimes1mx = _VIP[1] * onemx;
+  Real expetatimes1mx = exp(etatimes1mx);
+
+  for (int ind = maxind; ind >= 2; ind--) {              //_d2tod40[0]=d2
+    sumP = _d2tod40[pref0vp + ind] + onemx * sumP;       //_d2tod40[38]=d40
+    sumB = _d2tod40[pref0vp + ind] * ind + onemx * sumB; //_d2tod40[-2+40]=d40
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;          //_VIP[42]=f40
+  }                                                      //_VIP[2]=f0
+  for (int ind = 1; ind >= 0; ind--) {
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;
+  }
+  sumP = 1.0 + sumP * pow(onemx, 2.0);
```

Comment:
see above, replacing with multiply is recommended as it is much faster and potentially more accurate.